### PR TITLE
Issue 18358 object type and id for activities

### LINKED ISSUE #18358: Changing activity related interfaces for 8.2
# Extending the Activity Manager

The manager's `publishActivity()` method needs two new parameters $objectType (string) and $objectId (int),
which are used to filter the list of activities, e.g. in the right side bar, to a given object (file + id).
Possible solutions:
1. Don't change the interface
2. Change the interface

### Don't change the interface

Positive:
+ Interface is kept stable
+ Feature is still available

Negative:
- Interface does not match the signature of the implementation, this might be problematic when dealing with mocks in unit tests (which are generated from the interface):
```php
	public function publishActivity(..., $type, $priority);
```
vs.
```php
	public function publishActivity(..., $type, $priority, $objectType = '', $objectId = 0) { ... }
```

### Change the interface

Positive:
+ Interface matches the signature of the implementation
+ Feature is still available

Negative:
- Interface is broken :zap: - But let's be honest, there shouldn't be any other activity manager implementations out there (apart from unit tests, which mostly rely on the interface instead of the implementation, when they use a mock)

## Suggestion

My suggestion would be changing the interface, it keeps the code cleaner and the risk to break anything is quite low I'd say

---

# Extending the Activity Consumer

Similar to the manager, the consumer also needs the new arguments on the `receive()` method.
Possible solutions:
1. Don't change the interface (see above)
2. Change the interface (see above)
3. :new: Add a second interface `IConsumer2` or `IConsumerv2`

### Add a second interface

Positive:
+ Interface is kept stable
+ Feature is still available

Negative:
- Calling the consumers is more complicated:
```php
	public function publishActivity($app, $subject, $subjectParams, $message, $messageParams, $file, $link, $affectedUser, $type, $priority, $objectType = '', $objectId = 0) {
		foreach($this->getConsumers() as $c) {
			if ($c instanceof IConsumer) {
				$c->receive(..., $type, $priority);
			} else {
				$c->receive(..., $type, $priority, $objectType, $objectId);
			}
		}
	}
```

## Suggestion

My suggestion would be to use Solution 2 here again, as the only known IConsumer is again the one from the activity app itself, which would be updated immediately.

---

# Extending the Activity Extension

In case we change the IExtension interface, my suggestion would be to use the Solution 3. There are quite a number of extensions out there and we want people to keep using it. I have currently two changes on my mind, which make developing an extension easier for app devs. But for now they are not needed, just nice to have (opposed to the IConsumer and IManager changes which we need for the right sidebar feature)

@DeepDiver1975 awaiting your comment

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #18358 
Requires https://github.com/owncloud/activity/pull/363

Required for the right side bar feature of activity

### CODE REVIEW

**Comment by MorrisJobke** [comment]
```
09:46:12 PHP Fatal error:  Declaration of NoOpConsumer::receive() must be compatible with OCP\Activity\IConsumer::receive($app, $subject, $subjectParams, $message, $messageParams, $file, $link, $affectedUser, $type, $objectType, $objectId) in /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/activitymanager.php on line 369
09:46:12 PHP Stack trace:
09:46:12 PHP   1. {main}() /usr/local/bin/phpunit:0
09:46:12 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:537
09:46:12 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:105
09:46:12 PHP   4. PHPUnit_TextUI_Command->handleArguments() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:115
09:46:12 PHP   5. PHPUnit_Util_Configuration->getTestSuiteConfiguration() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:623
09:46:12 PHP   6. PHPUnit_Util_Configuration->getTestSuite() phar:///usr/local/bin/phpunit/phpunit/Util/Configuration.php:853
09:46:12 PHP   7. PHPUnit_Framework_TestSuite->addTestFiles() phar:///usr/local/bin/phpunit/phpunit/Util/Configuration.php:940
09:46:12 PHP   8. PHPUnit_Framework_TestSuite->addTestFile() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:405
09:46:12 PHP   9. PHPUnit_Util_Fileloader::checkAndLoad() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:333
09:46:12 PHP  10. PHPUnit_Util_Fileloader::load() phar:///usr/local/bin/phpunit/phpunit/Util/Fileloader.php:42
09:46:12 PHP  11. include_once() phar:///usr/local/bin/phpunit/phpunit/Util/Fileloader.php:58
```

File: lib/private/activity/event.php [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+use OCP\Activity\IEvent;
+use OCP\Activity\IManager;
+
+class Event implements IEvent {
+	/** @var IManager */
+	protected $manager;
+
+	/** @var string */
+	protected $user;
```
Comment (nickvergessen) [inline lib/private/activity/event.php]: those two should be removed, because unused

File: lib/private/activity/event.php [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OC\Activity;
+
+use OCP\Activity\IEvent;
+use OCP\Activity\IManager;
```
Comment (nickvergessen) [inline lib/private/activity/event.php]: same here

File: lib/private/activitymanager.php [hunk 3]
```diff
@@ -124,36 +127,72 @@ protected function getExtensions() {
...
-					$message,
-					$messageParams,
-					$file,
-					$link,
-					$affectedUser,
-					$type,
-					$priority);
-			} catch (\Exception $ex) {
-				// TODO: log the exception
+	public function generateEvent() {
```
Comment (PVince81) [inline lib/private/activitymanager.php]: Please expand the PHPDoc.
The purpose is to make it possible to create event objects to be used for publishing

File: lib/private/activitymanager.php [hunk 4]
```diff
@@ -124,36 +127,72 @@ protected function getExtensions() {
...
+	 * @param string $subject       A short description of the event
+	 * @param array  $subjectParams Array with parameters that are filled in the subject
+	 * @param string $message       A longer description of the event
+	 * @param array  $messageParams Array with parameters that are filled in the message
+	 * @param string $file          The file including path where this event is associated with
+	 * @param string $link          A link where this event is associated with
+	 * @param string $affectedUser  Recipient of the activity
+	 * @param string $type          Type of the notification
+	 * @param int    $priority      Priority of the notification
+	 * @return null
```
Comment (PVince81) [inline lib/private/activitymanager.php]: Deprecated ?

Comment (PVince81) [inline lib/private/activitymanager.php]: ah, maybe on the interface. Let me check it.

File: lib/public/activity/imanager.php [hunk 5]
```diff
@@ -38,22 +38,35 @@
...
+	 * @param string $message       A longer description of the event
+	 * @param array  $messageParams Array with parameters that are filled in the message
+	 * @param string $file          The file including path where this event is associated with
+	 * @param string $link          A link where this event is associated with
+	 * @param string $affectedUser  Recipient of the activity
+	 * @param string $type          Type of the notification
+	 * @param int    $priority      Priority of the notification
+	 * @return null
 	 * @since 6.0.0
+	 * @deprecated 8.2.0 Grab an IEvent from generateEvent() instead and use the publish() method
```
Comment (PVince81) [inline lib/public/activity/imanager.php]: aha, here it is :wink:

**Comment by PVince81** [comment]
- [ ] TEST: regular activities
- [ ] TEST: accepting federated share
- [ ] TEST: declining federated share

**Comment by PVince81** [comment]
- [ ] TEST: fileid in DB for regular user
- [ ] TEST: fileid in DB for shared file
- [ ] TEST: whether emails are still sent

(all tests done together with https://github.com/owncloud/activity/pull/363)

**Comment by PVince81** [comment]
- [ ] TEST: public link activities / fileid

**Comment by PVince81** [comment]
:+1:

File: lib/private/activitymanager.php (hunk 3 above)
Comment (nickvergessen) [inline lib/private/activitymanager.php]: done

**Comment by MorrisJobke** [comment]
> (all tests done together with owncloud/activity#363)

Dito. Did the same tests and it works :+1:

**Pull request merged**